pkg/disk/disk.go

```diff
@@ -0,0 +1,254 @@
+	})
+
+	log.Info("Deleting disk")
+
+	if req.DiskName == "" || req.DiskUuid == "" {
+		return &empty.Empty{}, grpcstatus.Error(grpccodes.InvalidArgument, "disk name and disk UUID are required")
+	}
+
+	if s.spdkClient != nil {
+		return &empty.Empty{}, s.spdkClient.DiskDelete(req.DiskName, req.DiskUuid)
+	}
+
+	return &empty.Empty{}, nil
+}
+
+func (s *Server) DiskGet(ctx context.Context, req *rpc.DiskGetRequest) (*rpc.Disk, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"diskType": req.DiskType,
+		"diskName": req.DiskName,
+		"diskPath": req.DiskPath,
+	})
+
+	log.Info("Getting disk info")
+
+	if req.DiskName == "" || req.DiskPath == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "disk name and disk path are required")
+	}
+
+	switch req.DiskType {
+	case rpc.DiskType_block:
+		ret, err := s.spdkClient.DiskGet(req.DiskName)
+		if err != nil {
+			return nil, grpcstatus.Error(grpccodes.Internal, err.Error())
+		}
+		return spdkDiskToDisk(ret), nil
+	case rpc.DiskType_filesystem:
+		// TODO: implement filesystem disk type
+		fallthrough
+	default:
+		return nil, grpcstatus.Errorf(grpccodes.Unimplemented, "unsupported disk type %v", req.DiskType)
+	}
+}
+
+func (s *Server) DiskReplicaInstanceList(ctx context.Context, req *rpc.DiskReplicaInstanceListRequest) (*rpc.DiskReplicaInstanceListResponse, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"diskType": req.DiskType,
+		"diskName": req.DiskName,
+	})
+
+	log.Info("Listing disk replica instances")
+
+	if req.DiskName == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "disk name is required")
+	}
+
+	switch req.DiskType {
+	case rpc.DiskType_block:
+		replicas, err := s.spdkClient.ReplicaList()
+		if err != nil {
+			return nil, grpcstatus.Error(grpccodes.Internal, err.Error())
+		}
+		instances := map[string]*rpc.ReplicaInstance{}
+		for name, replica := range replicas {
+			instances[name] = replicaToReplicaInstance(replica)
+		}
+		return &rpc.DiskReplicaInstanceListResponse{
+			ReplicaInstances: instances,
+		}, nil
+	case rpc.DiskType_filesystem:
+		// TODO: implement filesystem disk type
+		fallthrough
+	default:
+		return nil, grpcstatus.Errorf(grpccodes.Unimplemented, "unsupported disk type %v", req.DiskType)
+	}
+}
+
+func (s *Server) DiskReplicaInstanceDelete(ctx context.Context, req *rpc.DiskReplicaInstanceDeleteRequest) (*empty.Empty, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"diskType":            req.DiskType,
+		"diskName":            req.DiskName,
+		"diskUUID":            req.DiskUuid,
+		"replciaInstanceName": req.ReplciaInstanceName,
+	})
+
+	log.Info("Deleting disk replica instance")
+
+	if req.DiskName == "" || req.DiskUuid == "" || req.ReplciaInstanceName == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "disk name, disk UUID and replica instance name are required")
+	}
+
+	switch req.DiskType {
+	case rpc.DiskType_block:
+		err := s.spdkClient.ReplicaDelete(req.ReplciaInstanceName, true)
```

Comment:
The same question. Shouldn't disk params be used here for replica deletion? or not really required.

Reply:
We have a replicaMap (cache) where replica names server as keys, so no need to send disk name.